### PR TITLE
Feature/ne rasdaemon

### PR DESCRIPTION
**Pull request by tznind**
## Proposed Changes

This merges all changes in IsIdentifiableAsService and NERasdaemon into the extraction-refactoring branch.  

Changes include running IsIdentifiable as a service, custom rules and [outsourcing NER](https://github.com/SMI/SmiServices/tree/c3780f81c769fd7c33103fce726fb43a3897fc0a/src/microservices/Microservices.IsIdentifiable#rules)

## Types of changes

What types of changes does your code introduce? Tick all that apply.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [x] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation-Only Update (if none of the other choices apply)
  - In this case, ensure that the message of the head commit from the source branch is prefixed with `[skip ci]`

## Checklist

By opening this PR, I confirm that I have:

- [x] Reviewed the [contributing](https://github.com/SMI/SmiServices/blob/master/CONTRIBUTING.md) guidelines for this repository
- [x] Ensured that the PR branch is in sync with the target branch (i.e. it is automatically merge-able)
- [x] Updated any relevant API documentation
- [x] Created or updated any tests if relevant
- [x] Accurately updated the [CHANGELOG](https://github.com/SMI/SmiServices/blob/master/CHANGELOG.md)
  - NOTE: This ***must*** include any changes to any of the following files: default.yaml, any of the RabbitMQ server configurations, GlobalOptions.cs
- [x] Listed myself in the [CONTRIBUTORS](https://github.com/SMI/SmiServices/blob/master/CONTRIBUTORS.md) file 🚀
- [x] Requested a review by one of the repository maintainers

## Issues

If relevant, tag any issues that are *expected* to be resolved with this PR. E.g.:

- Closes #\<issue-number>
- ...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -48,6 +48,9 @@ jobs:
     - dotnet test ./tests/applications/Applications.DicomDirectoryProcessor.Tests/Applications.DicomDirectoryProcessor.Tests.csproj
 
     - ( cd data/stanford-ner && ./download.sh && cd - )
+    - ( cd data/tessdata && ./download.sh && cd - )
+    - sudo apt install tesseract-ocr
+    - sudo apt install libtesseract-dev
```
Comment (rkm) [inline .travis.yml]: Are these two still required?

**Comment by rkm** [comment]
Just scanning through this PR - do we run the nerd program as part of the test suite?

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,11 +14,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Added
```
Comment (rkm) [inline CHANGELOG.md]: Are we missing the nerd program from here?

**Comment by tznind** [comment]
> 
> 
> Just scanning through this PR - do we run the nerd program as part of the test suite?

Not currently.  There is a test on both sides for the communication protocol (i.e. parse string X) and there is a TCP level test in the Java.  

We have tested that it works locally, if we want to do integration tests I suggest we do it in the same place as the CohortExtractor=>CTPAnonymiser=>CohortPackager tests i.e. the integration project

File: src/common/Smi.Common/Messages/MessageHeader.cs [hunk 3]
```diff
@@ -73,7 +73,7 @@ public MessageHeader(IDictionary<string, object> encodedHeaders, Encoding enc)
             ProducerProcessID = (int)encodedHeaders["ProducerProcessID"];
             ProducerExecutableName = enc.GetString((byte[])encodedHeaders["ProducerExecutableName"]);
             Parents = GetGuidArrayFromEncodedHeader(encodedHeaders["Parents"], enc);
-            OriginalPublishTimestamp = (long)encodedHeaders["OriginalPublishTimestamp"];
+            OriginalPublishTimestamp = Convert.ToInt64(encodedHeaders["OriginalPublishTimestamp"]); // XXX error casting from Int32 to Int64 using (long)
```
Comment (rkm) [inline src/common/Smi.Common/Messages/MessageHeader.cs]: Can we clarify this change? The original data type is also a `long` from what I can see.

File: src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            // Deserialize the message from the delivery arguments
+            if (!SafeDeserializeToMessage(header, basicDeliverEventArgs, out ExtractFileStatusMessage message))
+                return;
+
+            // The path is taken from the message, however maybe it should be FileSystemOptions|ExtractRoot in default.yaml
+            // If the filename has a rooted path then the ExtractionDirectory is ignored by Path.Combine
+            var toProcess = new FileInfo( Path.Combine(message.ExtractionDirectory, message.AnonymisedFileName) );
+
+            if(!toProcess.Exists)
```
Comment (rkm) [inline src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs]: Do we really want to completely crash out here?

File: src/microservices/uk.ac.dundee.hic.nerd/src/uk/ac/dundee/hic/nerd/Program.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		try {
+			listener.close();
+		} catch (IOException e) {
+			// Ignore exceptions on close(), can't do anything about that anyway
+		}
+		shutdown=true;
+	}
+
+	Program() throws IOException, ClassCastException, ClassNotFoundException {
+		this.listener = new ServerSocket(1881,255,InetAddress.getByName("127.0.0.1"));
```
Comment (rkm) [inline src/microservices/uk.ac.dundee.hic.nerd/src/uk/ac/dundee/hic/nerd/Program.java]: Can we make this configurable in the config file? We may want to run across multiple hosts.

Comment (rkm) [reply src/microservices/uk.ac.dundee.hic.nerd/src/uk/ac/dundee/hic/nerd/Program.java]: Resolved offline

File: src/common/Smi.Common/Messages/MessageHeader.cs (hunk 3 above)
Comment (tznind) [reply src/common/Smi.Common/Messages/MessageHeader.cs]: In general Convert is more likely to work than a straight cast so if there was an Exception in it that is resolved by using Convert then its probably fine?

https://stackoverflow.com/questions/15394032/difference-between-casting-and-using-the-convert-to-method

Maybe if the timestamp is low Rabbit serializes it as number without Type?

File: src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs (hunk 4 above)
Comment (tznind) [reply src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs]: What would you like? Nack? or output message support for missing files? either would be fine by me

File: CHANGELOG.md (hunk 2 above)
Comment (tznind) [reply CHANGELOG.md]: I have added the description and template README for nerd https://github.com/SMI/SmiServices/commit/0691cbab955cf5652fe02b70d85e55c3ca60929e @jas88 can you fill in the readme to show how to configure/run it?

Thanks.

File: src/common/Smi.Common/Messages/MessageHeader.cs (hunk 3 above)
Comment (rkm) [reply src/common/Smi.Common/Messages/MessageHeader.cs]: Sure, my question was really "was this a change in response to a bug we found"? I'd prefer not to poke about with the internals of our messaging unless we've actually seen an issue.

File: src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs (hunk 4 above)
Comment (rkm) [reply src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs]: Just checking, but I think the CTP service handles a missing file by sending a "failure" message. We could do the same here?

This is hopefully not a common occurrence, but I'd rather not kill the service or leave the CohortPackager wondering why it's missing an image from its final accounting.

Comment (rkm) [reply src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs]: CTP currently does this: https://github.com/SMI/SmiServices/blob/d0c21df7f04c2d2365b6ab88ab09033352ced303/src/microservices/com.smi.microservices.ctpanonymiser/src/main/java/org/smi/ctpanonymiser/messaging/CTPAnonymiserConsumer.java#L73

Comment (rkm) [reply src/microservices/Microservices.IsIdentifiable/Service/IsIdentifiableQueueConsumer.cs]: Resolved offline

File: src/common/Smi.Common/Messages/MessageHeader.cs (hunk 3 above)
Comment (rkm) [reply src/common/Smi.Common/Messages/MessageHeader.cs]: Resolved offline. This was a change required to handle how Python3 serializes integer types.

**Pull request merged**